pkg/terraform/terraform.go

```diff
@@ -111,13 +113,143 @@ type TerraformEntrypoint struct {
+var moduleSourcePattern = regexp.MustCompile(`source\s*\=\s\"(.*)\"`)
+
 // NewTerraformClient creates a new Terraform client.
 func NewTerraformClient(workingDir string) *TerraformClient {
 	return &TerraformClient{
 		workingDir: workingDir,
 	}
 }
 
+// ModuleUsage locates all the usages of modules in all terraform entrypoints and vice versa.
+func ModuleUsage(ctx context.Context, rootDir string, skipUnresolvableModules bool) (*ModuleUsageGraph, error) {
+	entrypoints, err := GetEntrypointDirectories(rootDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get entrypoints: %w", err)
+	}
+	moduleUsages, err := modules(ctx, rootDir, skipUnresolvableModules)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get module usages: %w", err)
+	}
+	entrypointToModules := make(map[string]map[string]struct{})
+	modulesToEntrypoints := make(map[string]map[string]struct{})
+	for _, entrypoint := range entrypoints {
+		entrypointToModules[entrypoint.Path] = make(map[string]struct{})
+		recurseAndAppend(entrypoint.Path, entrypoint.Path, entrypointToModules, moduleUsages)
+	}
+	for _, modules := range entrypointToModules {
+		for module := range modules {
+			if _, ok := modulesToEntrypoints[module]; !ok {
+				modulesToEntrypoints[module] = make(map[string]struct{})
+			}
+		}
+	}
+	for module := range modulesToEntrypoints {
+		for entrypoint, modules := range entrypointToModules {
+			if _, modOK := modules[module]; modOK {
+				modulesToEntrypoints[module][entrypoint] = struct{}{}
+			}
+		}
+	}
+	return &ModuleUsageGraph{
+		EntrypointToModules:  entrypointToModules,
+		ModulesToEntrypoints: modulesToEntrypoints,
+	}, nil
+}
+
+func recurseAndAppend(rootPth, pth string, entrypointToModules map[string]map[string]struct{}, moduleUsages map[string]*Modules) {
+	if usage, ok := moduleUsages[pth]; ok {
+		for modulePath := range usage.ModulePaths {
+			entrypointToModules[rootPth][modulePath] = struct{}{}
+			recurseAndAppend(rootPth, modulePath, entrypointToModules, moduleUsages)
+		}
+	}
+}
+
+// Modules represents the details of the modules used by a given module or terraform entrypoint.
+type Modules struct {
+	ModulePaths        map[string]struct{}
+	ModuleOrEntrypoint string
+}
+
+// modules locates all terraform entrypoints or modules and finds all of their module usages.
+func modules(ctx context.Context, rootDir string, skipUnresolvableModules bool) (map[string]*Modules, error) {
+	logger := logging.FromContext(ctx)
+	matches := make(map[string]*Modules)
+	if err := filepath.WalkDir(rootDir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return fmt.Errorf("failed to walk directory %s: %w", path, err)
+		}
+
+		if d.IsDir() {
+			return nil
+		}
+
+		if filepath.Ext(path) != ".tf" {
+			return nil
+		}
+
+		file, err := os.Open(path)
+		if err != nil {
+			return fmt.Errorf("failed to open file %s: %w", path, err)
+		}
+
+		// Read the contents of the file
+		contents, err := io.ReadAll(file)
```

Comment:
This can be simplfiied a bit with os.Readfile(), which saves opening and closing a file handle.

Reply:
Done